src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp

```diff
@@ -70,560 +71,492 @@ enum SuperSpell
+                    return;
+
+                DoStartNoMovement(target);
+                ElementalFour->SetInCombatWithZone();
+                ElementalFour->CombatStart(target);
+                ElementalFour->SetFaction(me->GetFaction());
+                ElementalFour->SetUnitMovementFlags(MOVEMENTFLAG_ROOT);
+                ElementalFour->SetModifierValue(UNIT_MOD_RESISTANCE_FROST, BASE_VALUE, 0);
             }
 
-            DoZoneInCombat();
-        }
+            Talk(SAY_ELEMENTALS);
+        });
+    }
+
+    void KilledUnit(Unit* /*victim*/) override
+    {
+        Talk(SAY_KILL);
+    }
 
-        void FlameWreathEffect()
+    void TriggerArcaneCooldown()
+    {
+        scheduler.Schedule(5s, [this](TaskContext)
         {
-            std::vector<Unit*> targets;
-            ThreatContainer::StorageType const& t_list = me->GetThreatMgr().GetThreatList();
+            _arcaneCooledDown = true;
+        });
+    }
 
-            if (t_list.empty())
-                return;
+    void TriggerFireCooldown()
+    {
+        scheduler.Schedule(5s, [this](TaskContext)
+        {
+            _fireCooledDown = true;
+        });
+    }
 
-            //store the threat list in a different container
-            for (ThreatContainer::StorageType::const_iterator itr = t_list.begin(); itr != t_list.end(); ++itr)
-            {
-                Unit* target = ObjectAccessor::GetUnit(*me, (*itr)->getUnitGuid());
-                //only on alive players
-                if (target && target->IsAlive() && target->GetTypeId() == TYPEID_PLAYER)
-                    targets.push_back(target);
-            }
+    void TriggerFrostCooldown()
+    {
+        scheduler.Schedule(5s, [this](TaskContext)
+        {
+            _frostCooledDown = true;
+        });
+    }
 
-            //cut down to size if we have more than 3 targets
-            while (targets.size() > 3)
-                targets.erase(targets.begin() + rand() % targets.size());
+    void JustDied(Unit* /*killer*/) override
+    {
+        Talk(SAY_DEATH);
 
-            uint32 i = 0;
-            for (std::vector<Unit*>::const_iterator itr = targets.begin(); itr != targets.end(); ++itr)
-            {
-                if (*itr)
-                {
-                    FlameWreathTarget[i] = (*itr)->GetGUID();
-                    FWTargPosX[i] = (*itr)->GetPositionX();
-                    FWTargPosY[i] = (*itr)->GetPositionY();
-                    DoCast((*itr), SPELL_FLAME_WREATH, true);
-                    ++i;
-                }
-            }
-        }
+        instance->SetData(DATA_ARAN, DONE);
 
-        void UpdateAI(uint32 diff) override
+        if (GameObject* libraryDoor = instance->instance->GetGameObject(instance->GetGuidData(DATA_GO_LIBRARY_DOOR)))
         {
-            if (!UpdateVictim())
-                return;
+            libraryDoor->SetGoState(GO_STATE_ACTIVE);
+            libraryDoor->RemoveGameObjectFlag(GO_FLAG_NOT_SELECTABLE);
```

Comment:
I think it changes its status to locked, Karazhan doors have weird behaviors

Reply:
I agree. Also the issue with Malchezaar door